File: aisd/pracownie/1/chodnik-graf.py
n = 0
kostki = []

with open("200k.in") as fin:
    first = True
    for line in fin:
        if first == True:
            n = line.strip()
            first = False
        else:
            kostka = line.strip().split()            
            kostki.append((int(kostka[0]), int(kostka[1]), int(kostka[2])))




def zbuduj_graf(kostki):
    graf = {}
    for k in kostki:
        l, m, r = k
        if l not in graf:
            graf[l] = []
        graf[l].append(k)
    return graf


def dfs(lacznik, graf, chodnik):
    if chodnik and lacznik == 0:
        return chodnik
    
    if lacznik not in graf or len(graf[lacznik]) == 0:
        return None
    
    for k in list(graf[lacznik]):
        graf[lacznik].remove(k)

        wynik = dfs(k[2], graf, chodnik + [k])
        if wynik is not None:
            return wynik
        
        graf[lacznik].append(k)

    return None


def dfs_iterative(graf):
    stack = [(0, [], {k: v.copy() for k,v in graf.items()})]
    while stack:
        lacznik, chodnik, graph = stack.pop()

        if chodnik and lacznik == 0:
            return chodnik
        
        if lacznik not in graph or not graph[lacznik]:
            continue

        for k in list(graph[lacznik]):
            new_graph = { node: edges.copy() for node,edges in graph.items()}
            new_graph[lacznik].remove(k)
            stack.append((k[2], chodnik + [k], new_graph))
    return None
        



print(dfs_iterative(zbuduj_graf(kostki)))

File: aisd/pracownie/1/chodnik-graf.cpp
#include <iostream>
#include <vector>
#include <stack>
#include <unordered_map>

using namespace std;

struct Kostka {
    int l;
    int m;
    int r;
};

// stan dfsa
struct State {
    int current; //aktualny wierzcholek (czyli lacznik)
    vector<Kostka> path; //chodnik
    unordered_map<int, vector<Kostka> > graph_state;
};

bool znajdzChodnikDFS() {
    return true;
}


int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    cin >> n;
    vector<Kostka> wwszystkie_kostki(n);

    for (int i = 0; i < n; i++) {
        int l, m, r;
        cin >> l >> m >> r;
        Kostka k = {l, m, r};
        wwszystkie_kostki[i] = k;
    }

    unordered_map<int, vector<Kostka> > graf;

    for (const auto &k : wwszystkie_kostki) {
        graf[k.l].push_back(k);
    }

    stack<State> s;

    State f = {0, vector<Kostka>(), graf};
    s.push(f);
    bool found = false;
    vector<Kostka> res;

    while (!s.empty()) {
        State current_st = s.top();
        s.pop();

        if (!current_st.path.empty() && current_st.current == 0) {
            res = current_st.path;
            found = true;
            break;
        }

        if (current_st.graph_state.find(current_st.current) == current_st.graph_state.end()
            || current_st.graph_state[current_st.current].empty()
        ) {
            continue;
        }


        vector<Kostka> edges = current_st.graph_state[current_st.current];

        for (auto i = 0; i < edges.size(); i++) {
            Kostka k = edges[i];

            auto new_graph = current_st.graph_state;
            auto &edge_lst = new_graph[current_st.current];
            edge_lst.erase(edge_lst.begin() + i);

            vector<Kostka> new_path = current_st.path;
            new_path.push_back(k);

            State new_st = { k.r, new_path, new_graph };
            s.push(new_st); 
        }
    }

    if (found) {
        cout << res.size() << endl;

        for (auto &k : res) {
            cout << k.l << " " << k.m << " " << k.r << endl;
        }
    } else {
        cout << "BRAK";
    }
    return 0;
}
